x/mongo/driver/auth/oidc.go

```diff
@@ -0,0 +1,338 @@
+		return oa.accessToken, nil
+	}
+
+	cred, err := callback(ctx, args)
+	if err != nil {
+		return "", err
+	}
+
+	oa.accessToken = cred.AccessToken
+	oa.tokenGenID++
+	conn.SetOIDCTokenGenID(oa.tokenGenID)
+	if cred.RefreshToken != nil {
+		oa.refreshToken = cred.RefreshToken
+	}
+	return cred.AccessToken, nil
+}
+
+// TODO GODRIVER-3246: Implement OIDC human flow
+// This should only be called with the Mutex held.
+//func (oa *OIDCAuthenticator) getAccessTokenWithRefresh(
+//	ctx context.Context,
+//	callback OIDCCallback,
+//	refreshToken string,
+//) (string, error) {
+//
+//	cred, err := callback(ctx, &OIDCArgs{
+//		Version:      apiVersion,
+//		IDPInfo:      oa.idpInfo,
+//		RefreshToken: &refreshToken,
+//	})
+//	if err != nil {
+//		return "", err
+//	}
+//
+//	oa.accessToken = cred.AccessToken
+//	oa.tokenGenID++
+//	oa.cfg.Connection.SetOIDCTokenGenID(oa.tokenGenID)
+//	return cred.AccessToken, nil
+//}
+
+// invalidateAccessToken invalidates the access token, if the force flag is set to true (which is
+// only on a Reauth call) or if the tokenGenID of the connection is greater than or equal to the
+// tokenGenID of the OIDCAuthenticator. It should never actually be greater than, but only equal,
+// but this is a safety check, since extra invalidation is only a performance impact, not a
+// correctness impact.
+func (oa *OIDCAuthenticator) invalidateAccessToken(conn driver.Connection) {
+	oa.mu.Lock()
+	defer oa.mu.Unlock()
+	tokenGenID := conn.OIDCTokenGenID()
+	// If the connection used in a Reauth is a new connection it will not have a correct tokenGenID,
+	// it will instead be set to 0. In the absence of information, the only safe thing to do is to
+	// invalidate the cached accessToken.
+	if tokenGenID == 0 || tokenGenID >= oa.tokenGenID {
+		oa.accessToken = ""
+		conn.SetOIDCTokenGenID(0)
+	}
+}
+
+// Reauth reauthenticates the connection when the server returns a 391 code. Reauth is part of the
+// driver.Authenticator interface.
+func (oa *OIDCAuthenticator) Reauth(ctx context.Context, cfg *Config) error {
+	oa.invalidateAccessToken(cfg.Connection)
+	// The HTTPClient argument of the cfg will be nil on a Reauth call, so we populate
+	// it from the one stored in the Authenticator at Auth time, since the HTTPClient is only
+	// configured on driver startup. The HTTPClient will be needed for builtin provider callbacks
+	cfg.HTTPClient = oa.httpClient
+	// it should be impossible to get a Reauth when an Auth has never occurred,
+	// so we assume cfg was properly set. There is nothing to enforce this, however,
+	// other than the current driver code flow. If cfg is nil, Auth will return an error.
+	return oa.Auth(ctx, cfg)
+}
+
+// Auth authenticates the connection.
+func (oa *OIDCAuthenticator) Auth(ctx context.Context, cfg *Config) error {
+	var err error
+
+	if cfg == nil {
+		return newAuthError(fmt.Sprintf("config must be set for %q authentication", MongoDBOIDC), nil)
+	}
+	oa.httpClient = cfg.HTTPClient
```

Comment:
There's a data race detected here when running the OIDC tests. If possible, we should pass in the HTTP client when we create the authenticator at initialization rather than when we call `Auth` to avoid having to write to this field at run time.

See [data-race-1.txt](https://github.com/user-attachments/files/16057070/data-race-1.txt) for details.